ta/pkcs11/src/pkcs11_attributes.c

```diff
@@ -724,24 +724,18 @@ create_attributes_from_template(struct obj_attrs **out, void *template,
 			type = PKCS11_CKK_AES;
 			break;
 		default:
```

Comment:
What we are missing here is of course asymmetric keys support.

In that we have PKCS11_CKO_PUBLIC_KEY or PKCS11_CKO_PRIVATE_KEY depending which template is being processed.

For this reason I also added template_class to method signature so that detail can be carried to path where it gets added to actual object.

Reply:
One could probably implement it in here if missing then set class = template_class; and then carry on to sanitize_client_object.

Reply:
I agree with you. 

As we add GenerateKeyPair in the picture, we can add the template_class as you have suggested in create_attributes_from_template() and further based on the mecha, class can be template class and type would come from mecha.

case PKCS11_CKM_EC_KEY_PAIR_GEN:
			class = template_class;
			type = PKCS11_CKK_EC;
			break;
case PKCS11_CKM_RSA_PKCS_KEY_PAIR_GEN:
			class = template_class;
			type = PKCS11_CKK_RSA;
			break;

Reply:
Added class_id as unused param in latest PR.